### PR TITLE
Valleys C++ Mapgen

### PR DESCRIPTION
**Pull request by duane-r**
This is a mapgen based on the Valleys Mapgen mod. It features realistic terrain with rivers and an interconnected cave system.

See also: https://forum.minetest.net/viewtopic.php?f=9&t=11430

### CODE REVIEW

**Comment by Fixer-007** [comment]
+1000
I've played lua version and it was so beautiful!

**Comment by paramat** [comment]
Will review.

**Comment by paramat** [comment]
Generally many lines need shortening, no longer than 89 columns when using 4-space tabs.
Exceptions are tidy blocks that would look horrible broken up, such as the noise definitions in the constructor and the 3 'MapgenV7Params' functions.

I like the way dirt is placed on the cave floor if air is above.

**Comment by duane-r** [comment]
I think that's got the spacing right.

>I like the way dirt is placed on the cave floor if air is above. Can i use this idea in other core mapgens?

Of course.

**Comment by paramat** [comment]
Thanks could you squash into 1 commit? Makes it easier to do line comments. Then i'll go through it again.

**Comment by paramat** [comment]
Concerning function names.

Mgv7 only uses the words 'base terrain' to distinguish that from 'mountain terrain' (Mgv5 uses 'base terrain' incorrectly and i'll fix that).
Here the use of 'base terrain' alongside 'terrain' is confusing, suggesting different things.
'getGroundLevelAtPoint()' has to be named so because it is accessed from outside, but for all other functions within a mapgen the word 'terrain' is better.
Any function getting ground level should have the word 'level' in it.
'calcBiomes' should be 'calculateBiomes' to match 'calculateNoise', this also helps to distinguish it from 'bmgr->calcBiomes()' which is external.

So please could the function names be:
```
	float terrainLevelAtPoint();
	float terrainLevelFromMap();
	void calculateBiomes();
	void calculateNoise();
	virtual int generateTerrain();
	float terrainLevelFromNoise();
	float adjustedTerrainLevelFromNoise();
```

**Comment by duane-r** [comment]
>Is 'Biome *getBiomeAtPoint()' actually needed? I don't know why mgv7 has it, most other mapgens don't, so you can delete that if it's not needed by the mapgen itself.

It's in emerge.h too, but never used. Maybe someone wanted it for a lua function that never appeared. /shrug

**Comment by paramat** [comment]
Added line comments.

**Comment by duane-r** [comment]
>In the 5 lines above, is the 'this->' before 'spflags' really needed? 'spflags' seems already usable on it's own because of this line: 'this->spflags = sp->spflags;'.

I used that as a cheap way to avoid triggering water placement if lava was already placed (since they use the same random number [sr]). If you like, I can use another variable.

**Comment by paramat** [comment]
Thanks for the updating work. Now i'll test it and profile it for speed.
Is the river noise 'spread' larger than the original lua version? i think it needs to be.

**Comment by duane-r** [comment]
>Is the river noise 'spread' larger than the original lua version?

No, but it's very easy to change. I thought we were trying to make it like the original. If we change the river noise, are we going to change the fill as well?

**Comment by paramat** [comment]
Sorry, i deleted those comments. Lets keep the default noise parameters the same as the original for the moment. As you say they are easily settable.

**Comment by paramat** [comment]
This block of code is missing from makeChunk:
```
	// Pre-conditions
	assert(data->vmanip);
	assert(data->nodedef);
	assert(data->blockpos_requested.X >= data->blockpos_min.X &&
		data->blockpos_requested.Y >= data->blockpos_min.Y &&
		data->blockpos_requested.Z >= data->blockpos_min.Z);
	assert(data->blockpos_requested.X <= data->blockpos_max.X &&
		data->blockpos_requested.Y <= data->blockpos_max.Y &&
		data->blockpos_requested.Z <= data->blockpos_max.Z);
```

**Comment by paramat** [comment]
Even without the 'fast' flag it's a fast mapgen.
I love the effect of altitude chill and humid rivers, i saw a cluster of jungletrees around a river in savanna, an oasis effect.

**Comment by paramat** [comment]
Why is 'calculateBiomes' a function inside the mapgen when the external version in mapgen.cpp is identical?

**Comment by duane-r** [comment]
>I love the effect of altitude chill and humid rivers

There's an additional decrease in humidity based on soil thickness/altitude in the original that I didn't include (maybe because it was added since then). I wonder if I should try to squeeze that in?

**Comment by duane-r** [comment]
I corrected a potential problem in humidityByTerrain, where I was modifying the humidity parameter. I don't think C++ allows that when you pass by value, but this is probably better form anyway.

**Comment by paramat** [comment]
Here's a big fat :+1:

**Comment by sfan5** [comment]
:+1:

**Comment by paramat** [comment]
There was a change to emerge.cpp but i'll attempt a merge soon.

File: src/mapgen_valleys.cpp [hunk 1]
```diff
@@ -0,0 +1,1014 @@
...
+//#include "assert.h"
+
+//#include "util/timetaker.h"
+//#include "profiler.h"
+
+
+//static Profiler mapgen_prof;
+//Profiler *mapgen_profiler = &mapgen_prof;
+
+FlagDesc flagdesc_mapgen_valleys[] = {
```
Comment (nerzhul) [inline src/mapgen_valleys.cpp]: please add static const

File: src/mapgen_valleys.cpp [hunk 2]
```diff
@@ -0,0 +1,1014 @@
...
+	}
+
+	return humidity;
+}
+
+
+int MapgenValleys::getGroundLevelAtPoint(v2s16 p)
+{
+	// Base terrain calculation
+	s16 y = terrainLevelAtPoint(p.X, p.Y);
```
Comment (nerzhul) [inline src/mapgen_valleys.cpp]: return directly result without creating a variable & set this function inline

File: src/mapgen_valleys.h [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+	~MapgenValleysParams() {}
+
+	void readParams(const Settings *settings);
+	void writeParams(Settings *settings) const;
+};
+
+class MapgenValleys : public Mapgen {
+public:
+	EmergeManager *m_emerge;
+	BiomeManager *bmgr;
```
Comment (nerzhul) [inline src/mapgen_valleys.h]: why everything is public ? please set private fields properly

**Comment by duane-r** [comment]
Like this? --^

>return directly result without creating a variable & set this function inline

**Comment by nerzhul** [comment]
@duane-r yes, but it wasn't on this function :). But after your refactoring seems go to make this inline

**Comment by duane-r** [comment]
I can make it static, but getFlagStrNoEx complains when I try to pass it a const.

>please add static const

**Comment by duane-r** [comment]
>yes, but it wasn't on this function

Oh! Now I get it.

File: src/mapgen_valleys.cpp (hunk 1 above)
Comment (paramat) [inline src/mapgen_valleys.cpp]: This is how mgv7 is though.

File: src/mapgen_valleys.h (hunk 3 above)
Comment (paramat) [inline src/mapgen_valleys.h]: See mgv7, i think it is the same as this.

Comment (duane-r) [inline src/mapgen_valleys.h]: Nothing actually calls most of the methods though, so I can't see any reason not to make them private.

Comment (paramat) [inline src/mapgen_valleys.h]: Okay.

**Comment by paramat** [comment]
Hmmmm approves too so i'll attempt a merge later.

**Comment by paramat** [comment]
Merge failed in emerge.cpp i suspected, due to mgflat being unhidden.
Please could you fix the whitespaces too?
```
Applying: Add Valleys mapgen.
/home/matstandard/minetest/.git/rebase-apply/patch:391: space before tab in indent.
	         	| MG_VALLEYS_HUMID_RIVERS | MG_VALLEYS_ALT_CHILL;
/home/matstandard/minetest/.git/rebase-apply/patch:581: trailing whitespace.
				node_min - v3s16(0, 1, 0), 
/home/matstandard/minetest/.git/rebase-apply/patch:582: trailing whitespace.
				node_max + v3s16(0, 1, 0), 
/home/matstandard/minetest/.git/rebase-apply/patch:583: trailing whitespace.
				full_node_min, 
/home/matstandard/minetest/.git/rebase-apply/patch:748: trailing whitespace.
		// The penultimate step builds up the heights, but we reduce it 
error: patch failed: src/emerge.cpp:108
error: src/emerge.cpp: patch does not apply
```

**Comment by nerzhul** [comment]
Wait for a second review from me.

**Comment by duane-r** [comment]
>Please could you fix the whitespaces too?

Where? I didn't see any problems in emerge.cpp.

>(Did you notice 1337 lines are added?)

Not anymore. Had to break a long line.  : p

**Comment by duane-r** [comment]
I merged the upstream commits, so there shouldn't be any issues with merging. It's ready for more review.

File: src/mapgen_valleys.cpp [hunk 4]
```diff
@@ -0,0 +1,1013 @@
...
+
+	heatmap = noise_heat->result;
+	humidmap = noise_humidity->result;
+}
+
+
+// This ugly function keeps us from having to maintain two similar sets of
+//  complicated code to determine ground level.
+float MapgenValleys::terrainLevelFromNoise(
+		s16 x,
```
Comment (nerzhul) [inline src/mapgen_valleys.cpp]: i think you should group all those parameters into a structure passed by reference if possible to reduce memory copies

File: src/mapgen_valleys.h [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+	content_t c_stone;
+	content_t c_water_source;
+
+	float terrainLevelAtPoint(s16 x, s16 z);
+
+	void calculateNoise();
+
+	virtual int generateTerrain();
+	float terrainLevelFromNoise(
+			s16 x,
```
Comment (nerzhul) [inline src/mapgen_valleys.h]: same for this function as the next function, maybe group those parameters into a structure

Comment (paramat) [inline src/mapgen_valleys.h]: No, this is how the other mapgens are, if we make this change we can do it for all mapgens at once in a later PR.

File: src/mapgen_valleys.cpp (hunk 4 above)
Comment (paramat) [inline src/mapgen_valleys.cpp]: Which parameters?

Comment (paramat) [inline src/mapgen_valleys.cpp]: Maybe in a later PR.

File: src/mapgen_valleys.h (hunk 5 above)
Comment (duane-r) [inline src/mapgen_valleys.h]: Oops. Didn't see your reply until just now. Anyway, it's done.

Comment (paramat) [inline src/mapgen_valleys.h]: Okay no problem.

**Comment by paramat** [comment]
Travis is misbehaving, ignore.

**Comment by paramat** [comment]
3c6b2ffb10e053ad0b585347275034c4d889530d

**Comment by paramat** [comment]
The whitespace errors are detailed in the comment above, in case you want to tidy these up in another commit.

**Comment by duane-r** [comment]
Thanks for all your help everyone, especially paramat.

.........

>The whitespace errors are detailed in the comment above, in case you want to tidy these up in another commit.

I think I'm missing something. The comment 3c6b2ff shows "0 comments" at the bottom, and I can't see any mention above that (even the older comments that I remember). I'll look for anything that doesn't match.

**Comment by nerzhul** [comment]
Paramat you merged without my approvement, i said i should review another time the C++ code and you merged. Don't fortet the rule, 2 +1.
Now @duane-r or @paramat should fix the problems on the code.
@paramat please fix mapgens to have private variable everytime it's possible first, same for functions, it's important.

**Comment by BlockMen** [comment]
@nerzhul There is something called reverting in those cases...

**Comment by nerzhul** [comment]
@BlockMen yes i know, but this PR is not problematic, it's design features it's why i don't revert. But this should be fixed ASAP by @paramat or @duane-r

**Comment by duane-r** [comment]
Just let me know what needs to be done. I think I've already covered everything paramat and nerzhul have suggested so far, except for some source formatting. Most of the members are private now, and I used a structure for that long list of parameters to two functions.

**Comment by paramat** [comment]
For the whitespace report see the text formatted as code in this comment above https://github.com/minetest/minetest/pull/3508#issuecomment-171169879

> Wait for a second review from me.

I did, you added 2 line comments a few thread comments after writing that, and duane-r fixed those issues, see comment above, what needs to be fixed?
I merged because it already has 2 approvals, mine and sfan5's, it was almost merged 2 days ago but the merge failed. BlockMen there is no justification for reverting here.

> please fix mapgens to have private variable everytime it's possible

Okay.

**Comment by duane-r** [comment]
>In case the text does not appear for you it was

It appeared. I was just spacing out. : )

All trailing whitespace is gone, along with a few extra spaces I found out of place. Will I need to submit a new PR for this and further changes?

**Comment by paramat** [comment]
Yes new PR.

**Comment by paramat** [comment]
@duane-r 2 people report spawning underwater.

**Comment by duane-r** [comment]
>2 people report spawning underwater.

Yes, people frequently spawn in rivers. As I recall, the main spawn routine just checks that they're at ground level (river bottom) and above sea level. I'm not sure what to do about that. I've spawned over a cave entrance before as well. Caves aren't figured into getGroundLevelAtPoint's calculation.

**Comment by duane-r** [comment]
>>2 people report spawning underwater.

>Yes, people frequently spawn in rivers. As I recall, the main spawn routine just checks that they're at ground level (river bottom) and above sea level. I'm not sure what to do about that. I've spawned over a cave entrance before as well. Caves aren't figured into getGroundLevelAtPoint's calculation.

findSpawnPos in server.cpp checks to make sure there is air where the player spawns, but it also considers CONTENT_IGNORE to be air. In the cases where I'm spawning in rivers, the map is returning CONTENT_IGNORE. The most sensible answer to me would be to only allow air, but apparently the map object never returns air, forcing a spawn at 0,0,0.

The code in map.cpp is a bit over my head, so I'm not sure why it's not returning useful data.

**Comment by paramat** [comment]
Okay i'm working on spawn search currently so will work on it.
For mgwatershed i check river noise to avoid a spawn in a river https://github.com/paramat/minetest/blob/rivergen/src/mapgen_watershed.cpp#L229 the same thiing may be possible here, will look into it.

**Comment by duane-r** [comment]
>MAX_MAP_GENERATION_LIMIT

Nice! That makes it easy. Thanks.

**Pull request closed without merging**